sphinx_immaterial/custom_admonitions.py

```diff
@@ -0,0 +1,354 @@
+"""This module inherits from the generic ``admonition`` directive and makes the
+title optional."""
+from abc import ABC
+from pathlib import Path, PurePath
+import re
+from typing import List, Dict, Any, Tuple, Optional, Type, Union
+from docutils import nodes
+from docutils.parsers.rst import directives, Directive
+import jinja2
+import pydantic
+import sphinx.addnodes
+from sphinx.application import Sphinx
+from sphinx.config import Config
+from sphinx.environment import BuildEnvironment
+import sphinx.ext.todo
+from sphinx.locale import admonitionlabels, _
+from sphinx.util.logging import getLogger
+from sphinx.writers.html5 import HTML5Translator
+from .inline_icons import load_svg_into_builder_env
+
+logger = getLogger(__name__)
+
+# treat the todo directive from the sphinx extension as a built-in directive
+admonitionlabels["todo"] = _("Todo")
+
+
+class CustomAdmonitionConfig(pydantic.BaseModel):
```

Comment:
I am not too familiar with pydantic, but when I have used it, I have used `pydantic.dataclasses.dataclass` rather than `BaseModel` --- but perhaps you have a reason for preferring it this way.  Also, is it necessary to define `__init__` rather than relying on the built-in definition?  Is that needed to show the default values in autodoc?

Reply:
I was just following the pydantic docs about using the BaseModel class; I have no preference either way.

`__init__()` is needed to ensure the color components range is [0, 255], adopt the `name` as a `title` if not provided by the user, and yes the default values only show when documenting the params. If I just document the class's attributes, then autodoc doesn't show the default values - this may lead to confusion about what fields are required in the config's dict.

Reply:
I'm also using `__init__()` to limit the allowed `name` value to consist of only characters in the pattern `[a-zA-Z0-9\-_]`. Then the title can be extracted from the name with `" ".join(re.split(r"\W+", name)).title()`

Reply:
It looks like pydantic has a built-in validation mechanism: https://pydantic-docs.helpmanual.io/usage/validators/

Reply:
Oh cool! I'll try that out. I'm not a big fan of the way the pydantic docs are written (lead by example then list what wasn't shown in the example).

Reply:
Done. I also moved to documenting the class attributes now that all are optional except the `name` field.